pinot-common/src/main/java/org/apache/pinot/common/utils/request/RequestUtils.java

```diff
@@ -664,4 +670,49 @@ public static void applyTimestampIndexOverrideHints(
       }
     }
   }
+
+  /**
+   * Infers the expression type by recursively traversing the expression tree for function calls. The provided schema
+   * is used to resolve the types for identifier expressions. Note that for function calls, only scalar functions are
+   * supported here currently. Transform functions that don't have equivalent scalar functions, and aggregation
+   * functions aren't supported and {@link ColumnDataType#UNKNOWN} will be returned for them.
+   */
+  @Nullable
+  public static ColumnDataType inferExpressionType(@Nullable Expression expression, Schema schema) {
+    if (expression == null) {
+      return null;
+    }
+
+    if (expression.isSetIdentifier()) {
+      String columnName = expression.getIdentifier().getName();
+      FieldSpec fieldSpec = schema.getFieldSpecFor(columnName);
+      if (fieldSpec == null) {
+        return ColumnDataType.UNKNOWN;
+      }
+      return ColumnDataType.fromDataType(fieldSpec.getDataType(), fieldSpec.isSingleValueField());
+    }
+
+    if (expression.isSetLiteral()) {
+      LiteralContext literalContext = new LiteralContext(expression.getLiteral());
+      return ColumnDataType.fromDataType(literalContext.getType(), literalContext.isSingleValue());
+    }
+
+    if (expression.isSetFunctionCall()) {
+      Function fn = expression.getFunctionCall();
+      int numOperands = fn.getOperandsSize();
+      ColumnDataType[] argTypes = new ColumnDataType[numOperands];
+      for (int i = 0; i < numOperands; i++) {
+        ColumnDataType argType = inferExpressionType(fn.getOperands().get(i), schema);
+        argTypes[i] = argType != null ? argType : ColumnDataType.UNKNOWN;
+      }
+      FunctionInfo functionInfo = FunctionRegistry.lookupFunctionInfo(fn.getOperator(), argTypes);
+      if (functionInfo != null) {
+        Class<?> returnClass = functionInfo.getMethod().getReturnType();
```

Comment:
This is probably not enough, especially when a function returns `Object` because various types of value can be returned by the same function based on the argument, e.g. `JSON_EXTRACT_SCALAR`.
Ideally we want to use `SqlReturnTypeInference` to inference the return type of a function. This info should be available from `PinotOperatorTable`. All the working functions (at least in MSE) are already registered there

Reply:
Yeah that's a good point.

I've given the Calcite based approach a go, but I'm not sure if it's robust enough to cover all cases TBH. Also, there are gonna be type differences between MSE and SSE especially for the Calcite standard operators - for instance, in SSE `abs` always returns `DOUBLE` whereas the standard Calcite `ABS` operator returns the same type as the operand (so `INT` for `INT` etc.).